src/main/java/com/facebook/presto/pinot/grpc/PinotStreamingQueryClient.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot.grpc;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Grpc based Pinot query client.
+ */
+public class PinotStreamingQueryClient
+{
+    private final Map<String, GrpcQueryClient> grpcQueryClientMap = new HashMap<>();
+    private final Config config;
+
+    public PinotStreamingQueryClient(Config config)
+    {
+        this.config = config;
+    }
+
+    public Iterator<ServerResponseWrapper> submit(String host, int port, GrpcRequestBuilder requestBuilder)
+    {
+        GrpcQueryClient client = getOrCreateGrpcQueryClient(host, port);
+        return client.submit(requestBuilder);
+    }
+
+    private GrpcQueryClient getOrCreateGrpcQueryClient(String host, int port)
+    {
+        String key = String.format("%s_%d", host, port);
+        if (!grpcQueryClientMap.containsKey(key)) {
+            grpcQueryClientMap.put(key, new GrpcQueryClient(host, port, config.getMaxInboundMessageSize()));
+        }
+        return grpcQueryClientMap.get(key);
+    }
+
+    public static class Config
+    {
+        // Default max message size to 128MB
```

Comment:
The purpose of streaming connector is to resolve the issue that for large data scanning would put big memory pressure in pinot-server since, it requires to serialize all the results into a data table then stream it back.

1. This is only for segment level query, so, no aggregation is used here.
2. Pinot server will split the response into chunks and internally it will call next block to compute the results for next chunk, the results are collected and send back.
3. no compression as for now, I feel we can support it later on but requires further upgrade for the client lib
4. 128mb is picked for optimizing the batch size as typical pinot segment size is between 100MB to 1GB.

Reply:
Great description.. can you add this to the Pr description please